### PR TITLE
Feature Token binding for OAuth2.0 

### PR DESCRIPTION
**Pull request by inthirakumaaran**
### Proposed changes in this pull request

- Implementation of Token Binding for Oauth2.0

### Follow up actions

- trigger a migration in DB. Need to add new column TB_MANDATORY in IDN_OAUTH_CONSUMER_APPS

#### General

- [x] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [x] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [x] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/tokenBinding/TokenBindingHandler.java [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+/**
+ * This class handles Token Binding For oauth2.0 requests.
+ */
+public class TokenBindingHandler implements TokenBinding {
+    private static Log log = LogFactory.getLog(TokenBindingHandler.class);
+    private static String delimiter = "&#%";
+
+    private boolean tBSupportExist = false;
+
+    public void settBSupportEnabled(boolean tBSupportEnabled) {
```
Comment (denuwanthi) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/tokenBinding/TokenBindingHandler.java]: This method is not in proper camel case format.

File: components/org.wso2.carbon.identity.oauth.ui/src/main/resources/web/oauth/add-finish-ajaxprocessor.jsp [hunk 2]
```diff
@@ -54,13 +54,17 @@
 
 	boolean pkceMandatory = false;
 	boolean pkceSupportPlain = false;
+    boolean tokenBinding = false;
```
Comment (denuwanthi) [inline components/org.wso2.carbon.identity.oauth.ui/src/main/resources/web/oauth/add-finish-ajaxprocessor.jsp]: need formatting

File: components/org.wso2.carbon.identity.oauth.ui/src/main/resources/web/oauth/add-finish-ajaxprocessor.jsp [hunk 3]
```diff
@@ -54,13 +54,17 @@
...
 
 	if(request.getParameter("pkce") != null) {
 		pkceMandatory = true;
 	}
 	if(request.getParameter("pkce_plain") != null) {
 		pkceSupportPlain = true;
 	}
+    if(request.getParameter("tb") !=null){
```
Comment (denuwanthi) [inline components/org.wso2.carbon.identity.oauth.ui/src/main/resources/web/oauth/add-finish-ajaxprocessor.jsp]: need formatting

Comment (inthirakumaaran) [reply components/org.wso2.carbon.identity.oauth.ui/src/main/resources/web/oauth/add-finish-ajaxprocessor.jsp]: okay

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/tokenBinding/TokenBindingHandler.java (hunk 1 above)
Comment (inthirakumaaran) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/tokenBinding/TokenBindingHandler.java]: okay I ll change it

File: components/org.wso2.carbon.identity.oauth.stub/src/main/resources/OAuthAdminService.wsdl [hunk 4]
```diff
@@ -940,13 +941,13 @@
...
-            <soap:address location="https://localhost:9443/services/OAuthAdminService.OAuthAdminServiceHttpsSoap11Endpoint/"/>
+            <soap:address location="https://localhost:9444/services/OAuthAdminService.OAuthAdminServiceHttpsSoap11Endpoint/"/>
         </wsdl:port>
         <wsdl:port name="OAuthAdminServiceHttpsSoap12Endpoint" binding="tns:OAuthAdminServiceSoap12Binding">
-            <soap12:address location="https://localhost:9443/services/OAuthAdminService.OAuthAdminServiceHttpsSoap12Endpoint/"/>
+            <soap12:address location="https://localhost:9444/services/OAuthAdminService.OAuthAdminServiceHttpsSoap12Endpoint/"/>
         </wsdl:port>
         <wsdl:port name="OAuthAdminServiceHttpsEndpoint" binding="tns:OAuthAdminServiceHttpBinding">
-            <http:address location="https://localhost:9443/services/OAuthAdminService.OAuthAdminServiceHttpsEndpoint/"/>
+            <http:address location="https://localhost:9444/services/OAuthAdminService.OAuthAdminServiceHttpsEndpoint/"/>
```
Comment (ruwanta) [inline components/org.wso2.carbon.identity.oauth.stub/src/main/resources/OAuthAdminService.wsdl]: Please revert the unwanted changes like this

**Pull request merged**